### PR TITLE
Add function calling llama 2 7b

### PR DESCRIPTION
**Pull request by Fran-98**
Hi! I added the config and prompt formatting to run the function calling llama 2 model from Trelis. Also added the corresponding tutorial

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Looks great @Fran-98! Can you add an entry to the README?

**Comment by Fran-98** [comment]
@lantiga going to add a soon as I get to my pc

**Comment by Fran-98** [comment]
@lantiga Done!

File: chat/base.py [hunk 1]
```diff
@@ -262,6 +262,35 @@ def prompt_config(checkpoint_dir: Path, tokenizer: Tokenizer) -> Tuple[str, Tupl
...
         stop_tokens = ([tokenizer.eos_id],)
         return system_prompt, stop_tokens
+    
+    if re.search("Llama-2-7b-chat-hf-function-calling-v2", checkpoint_name):
+        # Has to be before the llama config
+        b_func, e_func = "<FUNCTIONS>", "</FUNCTIONS>\n\n"
+        b_inst, e_inst = "[INST]", "[/INST]"
+        b_sys, e_sys = "<<SYS>>\n", "\n<</SYS>>\n\n"
+        # This is an example for how to format functions for the model
+        from json import dumps
```
Comment (carmocca) [inline chat/base.py:272]: Can you make this a global import?

File: chat/base.py [hunk 2]
```diff
@@ -262,6 +262,35 @@ def prompt_config(checkpoint_dir: Path, tokenizer: Tokenizer) -> Tuple[str, Tupl
...
+                                    "description": "The search query string"
+                                }
+                            ]
+                        }
+        
+        system_prompt = (
+            "You are a helpful, respectful and honest assistant. Always answer as helpfully as"
+            "possible. Your only response should be JSON formatted functions"
+        )
+        function_list = dumps(function_metadata, indent=4).replace('{', '{{').replace('}', '}}') # Have to replace the curly braces to double curly braces to escape them
```
Comment (carmocca) [inline chat/base.py:289]: Is indenting relevant to the query?

File: tutorials/download_function_calling_llama_2.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+Function calling Llama extends the hugging face Llama 2 models with function calling capabilities.
+The model responds with a structured json argument with the function name and arguments.
+
+
+In order to use the checkpoint, download the weights and convert the checkpoint to the lit-gpt format.
+
+```bash
+pip install huggingface_hub
+
+python scripts/download.py --repo_id Trelis/Llama-2-7b-chat-hf-function-calling-v2 --access_token your_hf_token --from_safetensors true
```
Comment (carmocca) [inline tutorials/download_function_calling_llama_2.md:14]: It doesn't seem like this model is gated

```suggestion
python scripts/download.py --repo_id Trelis/Llama-2-7b-chat-hf-function-calling-v2 --from_safetensors true
```

File: tutorials/download_function_calling_llama_2.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                    }
+    
+    system_prompt = (
+        "You are a helpful, respectful and honest assistant. Always answer as helpfully as"
+        "possible. Your only response should be JSON formatted functions"
+    )
+    function_list = dumps(function_metadata, indent=4).replace('{', '{{').replace('}', '}}') # Have to replace the curly braces to double curly braces to escape them
+    
+    system_prompt = f"{b_func}{function_list.strip()}{e_func}{b_inst}{b_sys}{system_prompt.strip()}{e_sys}{'{prompt}'}{e_inst}\n\n"
+```
```
Comment (carmocca) [inline tutorials/download_function_calling_llama_2.md:57]: Can you link to the actual source code instead of copying it here?

**Review by carmocca — COMMENTED** [review]
Hi @Fran-98, what's your motivation for adding this model?

I don't see the value of integrating it into the chat script because the function-calling capabilities still need the ability of making HTTP requests and perhaps be in a loop.

Adding the model configuration is fine, however, the current changes don't showcase how the model is meant to be used. That is, unless you intend to copy-paste the json responses out of the chat replies.

So I would like to hear first about how you plan to use this model.

File: chat/base.py (hunk 1 above)
Comment (Fran-98) [reply chat/base.py:272]: Sure! Done

File: chat/base.py (hunk 2 above)
Comment (Fran-98) [reply chat/base.py:289]: As far as I know, it shouldn't be, but I wanted the config to stay as close as possible to the example given on the model page on huggingface.

File: tutorials/download_function_calling_llama_2.md (hunk 3 above)
Comment (Fran-98) [reply tutorials/download_function_calling_llama_2.md:14]: Good catch, copy&paste error

File: tutorials/download_function_calling_llama_2.md (hunk 4 above)
Comment (Fran-98) [reply tutorials/download_function_calling_llama_2.md:57]: Sure

**Comment by Fran-98** [comment]
> Hi @Fran-98, what's your motivation for adding this model?
> 
> I don't see the value of integrating it into the chat script because the function-calling capabilities still need the ability of making HTTP requests and perhaps be in a loop.
> 
> Adding the model configuration is fine, however, the current changes don't showcase how the model is meant to be used. That is, unless you intend to copy-paste the json responses out of the chat replies.
> 
> So I would like to hear first about how you plan to use this model.

Hi @carmocca!
First of all thanks for the corrections, I'm going to add/change those asap.
My main motivation to get the model working is for testing purposes. I know that the model is intended to be used to call functions and be integrated with other systems, but it also works wonders as a way to get almost always a well formated JSON as a response, something that I was struggling with Llama 2 7b.
As for the chat integration, I though it could be useful for manually testing prompts, I doubt it could have another use besides that.
Feel free to ask anything more you need!

**Comment by carmocca** [comment]
Ah good, if that's the expectation then I'm on board

**Review by carmocca — APPROVED** [review]

**Pull request merged**